### ISSUE TITLE
Building in Windows

### ISSUE THREAD

**Issue by alexander-us85** [rpav/c2ffi#29]
I've successfully build llvm and then generated Visual Studio project files for c2ffi with CMake on Windows.
While building I got many errors like this one:
"error : unknown type name 'off_t'; did you mean '_off_t'?"

Is there any possibility to build c2ffi under windows or this is fundamentally impossible?

**Comment by rpav** [comment]
I don't think it should be impossible; if you've got clang, it should build.  I'm rather unfamiliar with Windows as a platform; there may be some odd differences that may need accounted for.  However, in this case, I don't think I use `off_t` at all, so this is probably something from clang or LLVM.

Are you using clang to compile or the Microsoft compiler?  You should if nothing else try compiling with clang, and make sure it's configured for C++11, since I believe the clang source is c++11 (and c2ffi will require it in the future).

**Comment by alexander-us85** [comment]
@rpav Thank you very much for the quick response. Yes, I've successfully built llvm and clang on windows, but I can't figure out the way to compile c2ffi using executables I've built. So I created a project for Visual Studio with cmake and then chose "Clang 3.7 with Microsoft CodeGen" as the platform toolset in project preferences. So, as I understood, I only can use microsoft's version of clang, not mine. It worth to mention that I turned on C++11 support in project preferences.

**Comment by emmatyping** [comment]
I was able to generate build files, but beyond some missing includes of clang files (which I can assume is just some extra customization on my part needed) it appears you are using getopt for the executable. getopt is not available on Windows, so some code changes would be needed.

**Comment by ptome** [comment]
I was able to successfully compile on Windows 10.  I found this issue when searching for a solution, so I'm posting my findings for others.  I tried building with Visual Studio and the LLVM extension (not the Microsoft clang compiler), but the binary distribution of LLVM does not provide everything needed.  I gave up on recompiling LLVM et al from source, maybe others can be successful using this path.

In the end, I used MSYS2/MingW64.  I installed the following (and dependencies), using `pacman`:
- `mingw64/mingw-w64-x86_64-gcc`
- `mingw64/mingw-w64-x86_64-cmake`
- every `LLVM` package under `mingw64/` 

In the `c2ffi` source folder, I edited `CMakeLists.txt` with two changes:
1. In the `target_link_libraries` list, I removed `LLVM` and added all the individual libraries, obtained from running `llvm-config --libs`
2. After these libraries, I added the library called `version`

I'm not sure why these edits are needed, it could be a problem with LLVM in MSYS2, I'm not knowledgeable enough.

Then, in the MSYS2 MinGW64 Shell, starting from the `c2ffi` source directory
```
mkdir build
cd build
cmake -G "MSYS Makefiles" ..
make
```
By default `cmake` will generate Visual Studio solution files on Windows, we need to tell it to generate Makefiles.  This will build `c2ffi` with `mingw64-gcc`.  I'm attaching my log files and `CMakeLists.txt` for reference.

If you want to try the Visual Studio route, I think it should be possible:

1. Compile LLVM and tools (clang, etc) from source (use Windows native cmake)
2. Install the [LLVM Compiler Toolchain extension](https://marketplace.visualstudio.com/items?itemName=LLVMExtensions.llvm-toolchain) and configure it to your LLVM installation
3. In the `c2ffi` source directory, use `cmake` with option `-T llvm` to generate solution files that will build with the `llvm` platform in Visual Studio.

I would like to know if anyone did this successfully using Visual Studio, since it would add other options for `c2ffi`.

[CMakeLists.txt](https://github.com/rpav/c2ffi/files/2884976/CMakeLists.txt)
[log.txt](https://github.com/rpav/c2ffi/files/2884986/log.txt)
